server/src/main/java/io/crate/session/Session.java

```diff
@@ -316,13 +322,17 @@ public CoordinatorSessionSettings sessionSettings() {
     }
 
     public void parse(String statementName, String query, List<DataType<?>> paramTypes) {
+        long before = System.currentTimeMillis();
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug("method=parse stmtName={} query={} paramTypes={}", statementName, query, paramTypes);
         }
 
         Statement statement;
         try {
             statement = SqlParser.createStatement(query);
+            long parseTime = System.currentTimeMillis() - before;
```

Comment:
I don't think it can run forever unless there is a bug in antlr. It might take long but we account for it. 

Most problematic part is a planner (https://github.com/crate/crate/issues/16066) but there we will check timeout after each X iterations (in a follow up).